fontrenderer.1
.TH fontrenderer 1 "" Jul-2012 ""
.\" manual page by Dominik Kaaser
.\" using pages made by Peter Palfrader as references
.\" This manual page is under the same license as FontRenderer itself. See below.

.SH NAME
.LP
FontRenderer \- Converts input text to polygonal data

.SH SYNOPSIS
.LP
\fBfontrenderer\fP [\fIoptions\fP] fontfile [\fIinput\fP [\fIoutput\fP]]

.SH DESCRIPTION
.LP
\fBFontRenderer\fR is a tool that converts a text to polygonal data in \*(lq.line\*(rq-format using the outline of a specified font.

.SS
OPTIONS
.IP "\fIinput-file\fP" 8
Is the file which contains the text to be rendered.
If not specified or a minus (\-) is given, the code will be read from STDIN.

.IP "\fIoutput-file\fP" 8
Is the file to write the polygonal data to.
If not specified or a minus (\-) is given,
the code will be written to STDOUT.


.IP "\fB\-l\fR, \fB--line-height\fP" 8
Set the line height relative to the character height extracted from the font face.
A value of 1.0 represents the default line height.

.IP "\fB\-w\fR, \fB--letter-spacing\fP" 8
Set the letter spacing relative to the character width extracted from the font face.
A value of 1.0 represents the default letter spacing.

.IP "\fB\-s\fR, \fB--sample-rate\fP" 8
Set the sample rate for Bezier curve sampling. The value represents a number smaller than the total number of vertices per 1.0 units arc length sampled on the Bezier curve.

.IP "\fB\-u\fR, \fB--no-scale\fP" 8
Do not scale and shift the resulting data to the unit square.

.IP "\fB\-h\fP, \fB--help\fP" 8
Print a help message to STDOUT and exit

.IP "\fB\-v\fP, \fB--version\fP" 8
Print version info and exit


.SH FONT FILES

The font file can be any file, that can be read and parsed by the FreeType 2 library. The font file must define a valid outline and thus a pixel based font map cannot be specified.

.SH OUTPUT FORMAT

The resulting polygonal data is written in \*(lq.line\*(rq-format. Multiple polygonal chains may be specified within one file. A polygonal chain is specified by first the number of vertices followed by a list of vertices, one vertex per line. The vertices are specified as pairs of floating-point numbers separated by at least one white space character.

.LP
Note, that polygonal chains are not automatically closed to form closed polygons. Thus, the first and last point of a polygonal chain have to match in order to regard the polygonal chain as a closed polygon.

.SH NOTES

.LP
The simplicity of the output is based on the definition of the font file and is neither checked nor guaranteed.

.SH BUGS

.LP
Please report bugs to fontrenderer@kaaser.at.
This program is still a beta release, so you should expect to find some.

Also have a look at the project's web-site, 
\fIhttp://font.kaaser.at/\fP.

.SH AUTHOR

.LP
Dominik Kaaser <dominik@kaaser.at>
.LP
See also the CONTRIBUTORS file in the source package.



src/fontrenderer.cpp
//============================================================================
// Name        : fontrenderer.cpp
// Author      : Dominik S. Kaaser
// Copyright   : (C) Copyright 2012
// Description : Font-Renderer
//============================================================================

#define VERSION "1.1.0"
#define NAME "FontRenderer"

#include <iostream>
#include <istream>
#include <fstream>
#include <getopt.h>
#include <memory>

#include "renderer.h"

using namespace dcm;
using namespace std;

static string programName;

static struct option long_options[] = {
		{ "help", no_argument, 0, 'h' },
		{ "version", no_argument, 0, 'v' },
		{ "line-height", required_argument, 0, 'l' },
		{ "letter-spacing", required_argument, 0, 'w' },
		{ "sample-rate", required_argument, 0, 's' },
		{ "no-scale", required_argument, 0, 'u' },
		{ NULL, 0, 0, 0 }
};

static string options_descriptions[] = {
		"show this help",
		"print version info and exit",
		"set line height",
		"set letter spacing",
		"set sample rate for Bezier curves",
		"do not scale to unit square",
		""
};

static string options_params[] = {
		"",
		"",
		"height",
		"width",
		"samples",
		"",
		""
};

static const int usage(const int error = 0) {

	ostream& out = error ? cerr : cout;

	out << NAME << ": Converts a text to polygonal data in line format using a" << endl
			<< "specified font." << endl
			<< endl
			<< "Usage: " << programName << " [options] fontfile [input [output]]" << endl
			<< endl
			<< "Options:" << endl;

	int i = 0;
	while (long_options[i].val) {
		out << "    -" << (char) long_options[i].val << " " << options_params[i] << endl << "    " << "--" << left
				<< setw(16) << long_options[i].name << options_descriptions[i] << endl << endl;
		i++;
	}
	return error;
}

int main(int argc, char** argv) {

	try {
		programName = string(argv[0]);

		Renderer* object = new Renderer();

		char* font;
		istream* input = &cin;
		ostream* output = &cout;
		auto_ptr<ifstream> fileInputStream;
		auto_ptr<ofstream> fileOutputStream;

		int opt;
		int index = 0;

		while ((opt = getopt_long(argc, argv, "hvl:w:s:u", long_options, &index)) != -1) {

			switch (opt) {
			case 'h':
				delete object;
				return usage();
			case 'u':
				object->setScaleToUnit(false);
				break;
			case 'l':
				object->setLineHeight(atof(optarg));
				break;
			case 'w':
				object->setLetterSpacing(atof(optarg));
				break;
			case 's':
				object->setSampleRate(atoi(optarg));
				break;
			case 'v':
				cout << NAME << " version " << VERSION << endl << "Copyright (C) 2012 Dominik S. Kaaser" << endl
						<< "http://font.kaaser.at - dominik@kaaser.at" << endl;
				return 0;
			case '?':
				delete object;
				return usage(1);
			default:
				abort();
				break;
			}
		}

		switch (argc - optind) {
		case 3:
			if (strcmp(argv[optind + 2], "-") != 0) {
				fileOutputStream = auto_ptr<ofstream>(new ofstream(argv[optind + 2], ios::out | ios::trunc));
				if (!fileOutputStream->is_open())
					throw(string("Could not open output file: ") + string(argv[optind + 2]));
				output = fileOutputStream.get();
			}
			/* no break */
		case 2:
			if (strcmp(argv[optind + 1], "-") != 0) {
				fileInputStream = auto_ptr<ifstream>(new ifstream(argv[optind + 1], ios::in));
				if (!fileInputStream->is_open())
					throw(string("Could not open input file: ") + string(argv[optind + 1]));
				input = fileInputStream.get();
			}
			/* no break */
		case 1:
			font = argv[optind];
			break;
		default:
			return usage(1);
		}

		object->loadFace(font);

		object->input(*input);

		object->output(*output);

		delete object;

	} catch (string &exception) {
		cerr << exception << endl;
		return 2;
	}

	return 0;
}
